ENH: Enable DataFrame.corrwith to compute rank correlations

dsaxton opened the pull request:
This PR is to enable `DataFrame.corrwith` to calculate rank correlations in addition to Pearson's correlation (and should hopefully be fully backwards compatible), as well as clarifying functionality in the docstring .  An error message is generated if the user specifies a form of correlation that isn't implemented.  Tests were added for the new behavior.

closes #22328 
closes #21925

Commit:
ENH: Enable corrwith to compute rank and callable correlation methods

jreback reviewed (changes requested):
since you are changing the impl can you run the asv for perf check

dsaxton commented:
I'm having some trouble finding a relevant benchmark for `corrwith` (a `grep` on the benchmarks folder returns no hits).  Should one be added to the `Correlation` class in `stats_ops.py`?

dsaxton commented:
Here is one I put together which seems to suggest around a 33-50% speed-up with the new implementation.

```python
import numpy as np
from pandas import DataFrame, Series

np.random.seed(2357)

def corrwith_old(df1, df2, axis=0, drop=False):
    axis = df1._get_axis_number(axis)
    this = df1._get_numeric_data()

    if isinstance(df2, Series):
        return this.apply(df2.corr, axis=axis)

    df2 = df2._get_numeric_data()

    left, right = this.align(df2, join='inner', copy=False)

    # mask missing values
    left = left + right * 0
    right = right + left * 0

    if axis == 1:
        left = left.T
        right = right.T

    # demeaned data
    ldem = left - left.mean()
    rdem = right - right.mean()

    num = (ldem * rdem).sum()
    dom = (left.count() - 1) * left.std() * right.std()

    correl = num / dom

    if not drop:
        raxis = 1 if axis == 0 else 0
        result_index = this._get_axis(raxis).union(df2._get_axis(raxis))
        correl = correl.reindex(result_index)

    return correl


def corrwith_new(df1, df2, axis=0, drop=False, method='pearson'):
    if method not in ['pearson', 'spearman', 'kendall']:
        raise ValueError("method must be either 'pearson', "
                         "'spearman', or 'kendall', '{method}' "
                         "was supplied".format(method=method))

    axis = df1._get_axis_number(axis)
    this = df1._get_numeric_data()

    if isinstance(df2, Series):
        return this.apply(lambda x: df2.corr(x, method=method),
                          axis=axis)

    df2 = df2._get_numeric_data()
    left, right = this.align(df2, join='inner', copy=False)

    if axis == 1:
        left = left.T
        right = right.T

    correl = (left.apply(lambda x:
                         x.corr(right[x.name], method=method)))

    if drop:
        correl.dropna(inplace=True)

    return correl


data1 = DataFrame(np.random.normal(size=(10**6, 10)))
data2 = DataFrame(np.random.normal(size=(10**6, 10)))

%timeit corrwith_old(data1, data2)
%timeit corrwith_new(data1, data2)
```

jreback commented:
can u add this as an asv

in particular i suspect you method will slow down when used with a larger number of columns 

so pls show that as well

dsaxton commented:
I've added the new benchmark function `time_corrwith` to `stats_ops.py` within the `Correlation` class.  I'll try to run the benchmark locally (although I botched the setup of `asv` on my main computer and it seems to be broken).

dsaxton commented:
I think your suspicion may have been accurate regarding the dimension of the inputs.  In any case, regarding the tests it looks like there's an issue with `scipy` during the travis build.  Am I missing an import somewhere perhaps?

WillAyd commented:
@dsaxton we test quite a few different configurations, not all of which have SciPy installed. If your tests depend on that package you should use the `skip_if_no_scipy` decorate which you'll see used in other places in that same module

mroeschke commented on asv_bench/benchmarks/stat_ops.py:
```diff
@@ -109,6 +109,10 @@ class Correlation(object):
         self.df.corr(method=method)
+
+    def time_corrwith(self, method):
+        self.df.corrwith(df2, method=method)
```
`df2` should be `self.df2` here.

dsaxton replied:
Good catch, thank you.

dsaxton commented:
@jreback If a certain benchmark is actually part of the PR, what would you say is the most straightforward way to show that there isn't a degradation in performance?  Here the `time_corrwith` function will not work for the current implementation because it uses all three correlation types.  I suppose it could temporarily be modified to look only at Pearson?

WillAyd commented:
You should have for all three. The two new ones will fail since they don't exist on master but that's OK - still sets a baseline going forward and we can still get insights out of the Pearson benchmark

dsaxton commented:
![screen shot 2018-08-16 at 4 31 22 pm](https://user-images.githubusercontent.com/2658661/44236661-643f9680-a173-11e8-941e-9ddb8ffd4447.png)

Here are the results I get running the `stat_ops.Correlation` benchmarks (all `DataFrames` involved are comprised of 1000 x 30 Gaussian arrays).  The Pearson calculation somehow took longer than that of `DataFrame.corr`, even though the latter should actually be doing more work since it computes the full 30 x 30 correlation matrix.  I'll need to run it on the current `corrwith` to see how that performs.

dsaxton commented:
I modified the body of the function to mimic how the current calculation is being done in the special case where `method=pearson` (in the event that there was somehow an overhead associated with repeatedly calling `Series.corr`) and I saw essentially the same performance as before (shown below).  My guess then is that `DataFrame.corr` gets its speed from `nancorr` within `pandas._libs.algos`.  Could using functions from `algos` inside `corrwith` as well make sense?

```python
        if method in ['spearman', 'kendall']:
            correl = (left.apply(lambda x:
                                 x.corr(right[x.name], method=method)))
        else:
            correl = (((right - right.mean()) * (left - left.mean())).mean()
                      / right.std() / left.std())
```

<img width="448" alt="screen shot 2018-08-16 at 6 52 16 pm" src="https://user-images.githubusercontent.com/2658661/44240919-6828e400-a186-11e8-9e75-42799f03273b.png">

jreback reviewed (changes requested):
pls add a whatsnew note in other enhancements

jreback commented on pandas/core/frame.py:
```diff
@@ -6655,6 +6655,11 @@ def corr(self, method='pearson', min_periods=1):
+
+        See Also
+        -------
+        DataFrame.corrwith
```
can you add Series.corr here as well

jreback reviewed (changes requested):
what do the asv's show ?

jreback reviewed (changes requested):
pls add a whatsnew as well (with both issues)

jreback commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -248,6 +248,31 @@ def test_corrwith_mixed_dtypes(self):
 
+    @td.skip_if_no_scipy
+    def test_corrwith_spearman(self):
+        # GH PR #22375
```
change these to the issue numbers, rather than the PR number

jreback commented:
don't post pictures of the asv's just copy-paste the text pls.

dsaxton commented:
`asv` results for the `Correlation` class (`DataFrames` are built from `np.random.randn(1000, 30)`):

```
> asv continuous -f 1.1 origin/corrwith-dev HEAD -b stat_ops.Correlation
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Building a511a498 for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt........................................
·· Installing into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 6 total benchmarks (2 commits * 1 environments * 3 benchmarks)
[  0.00%] · For pandas commit hash a511a498:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  0.00%] ··· Running (stat_ops.Correlation.time_corr--)...
[ 16.67%] ··· stat_ops.Correlation.time_corr                                                                                         ok
[ 16.67%] ··· ========== =============
                method                
              ---------- -------------
               spearman     93.6±2ms  
               kendall      219±4ms   
               pearson    2.14±0.09ms 
              ========== =============

[ 33.33%] ··· stat_ops.Correlation.time_corrwith_cols                                                                                ok
[ 33.33%] ··· ========== ============
                method               
              ---------- ------------
               spearman    51.8±1ms  
               kendall     20.6±1ms  
               pearson    17.0±0.7ms 
              ========== ============

[ 50.00%] ··· stat_ops.Correlation.time_corrwith_rows                                                                                ok
[ 50.00%] ··· ========== ==========
                method             
              ---------- ----------
               spearman   758±30ms 
               kendall    493±7ms  
               pearson    258±7ms  
              ========== ==========

[ 50.00%] · For pandas commit hash a511a498:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 50.00%] ··· Running (stat_ops.Correlation.time_corr--)...
[ 66.67%] ··· stat_ops.Correlation.time_corr                                                                                         ok
[ 66.67%] ··· ========== =============
                method                
              ---------- -------------
               spearman     101±6ms   
               kendall      226±9ms   
               pearson    2.16±0.03ms 
              ========== =============

[ 83.33%] ··· stat_ops.Correlation.time_corrwith_cols                                                                                ok
[ 83.33%] ··· ========== ============
                method               
              ---------- ------------
               spearman    52.1±4ms  
               kendall     21.1±2ms  
               pearson    16.5±0.6ms 
              ========== ============

[100.00%] ··· stat_ops.Correlation.time_corrwith_rows                                                                                ok
[100.00%] ··· ========== ==========
                method             
              ---------- ----------
               spearman   752±20ms 
               kendall    490±4ms  
               pearson    266±3ms  
              ========== ==========


BENCHMARKS NOT SIGNIFICANTLY CHANGED.
> 
```

TomAugspurger commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -182,6 +182,7 @@ Other Enhancements
 - :func:`to_timedelta` now supports iso-formated timedelta strings (:issue:`21877`)
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
+- :meth:`DataFrame.corrwith` now supports Spearman's rank correlation as well as Kendall's tau. (:issue:`21925`)
```
This makes is sound like Knedall's tau was the previously supported method. Maybe "... supports Spearman's rank correlation and Kendall's tau for `method`"

dsaxton replied:
Good point, will make that change.

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -6804,54 +6810,58 @@ def cov(self, min_periods=None):
-        num = (ldem * rdem).sum()
-        dom = (left.count() - 1) * left.std() * right.std()
+        correl = (left.apply(lambda x:
+                             x.corr(right[x.name], method=method)))
```
Does this break when left or right have duplicate columns? Can you add a test ensuring that it works? It seems like we support that on master.

dsaxton replied:
I'll test that out.  What would you say the expected behavior should be in this case?  Would it make sense to first drop duplicate columns from `self` and `other`?

dsaxton replied:
Yes, I believe this breaks with duplicate columns.  There might be a way to still rely on `apply`, but if not I'll try to think of something else.  In the worst case a `for` loop should work, but I'd like to avoid that if possible.

dsaxton replied:
@TomAugspurger I think this is actually an error on `master`, but I may have a fix.

```
import numpy as np
import pandas as pd

df1 = pd.DataFrame(np.random.normal(size=(10, 3)))
df2 = pd.DataFrame(np.random.normal(size=(10, 3)))
df2 = pd.concat((df2, pd.DataFrame(np.random.randn(10))), axis=1)

df1.corrwith(df2)
```

A possible fix within `corrwith` is to create `correl` by applying a `map` to the `zipped` columns

```
pd.Series(map(lambda x: pd.Series(x[0]).corr(pd.Series(x[1]), method="spearman"),
              zip(left.values.T, right.values.T)),
          index=left.columns)
```

I can commit this as a change if you think it makes sense.

TomAugspurger commented:
Duplicate column names doesn't necessarily mean duplicate values.

We handle this correctly on master, so you should be able to add a test on
master that passes, and ensure it still passes on your branch.

On Thu, Aug 23, 2018 at 7:49 AM Daniel Saxton <notifications@github.com>
wrote:

> *@dsaxton* commented on this pull request.
> ------------------------------
>
> In pandas/core/frame.py
> <https://github.com/pandas-dev/pandas/pull/22375#discussion_r212295410>:
>
> >          if axis == 1:
>              left = left.T
>              right = right.T
>
> -        # demeaned data
> -        ldem = left - left.mean()
> -        rdem = right - right.mean()
> -
> -        num = (ldem * rdem).sum()
> -        dom = (left.count() - 1) * left.std() * right.std()
> +        correl = (left.apply(lambda x:
> +                             x.corr(right[x.name], method=method)))
>
> I'll test that out. What would you say the expected behavior should be in
> this case? Would it make sense to first drop duplicate columns from self
> and other?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/22375#discussion_r212295410>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIoYTVuSsxAk89rEjHQ6FBOjIfF1aks5uTqTAgaJpZM4V_AKR>
> .
>

dsaxton commented:
@TomAugspurger I just pushed some changes to that included the above (`zip` / `map` instead of `apply`, more clear `whatsnew`), and also added a relatively simple test to check if the method can handle duplicate columns (below).  I may have mauled the `git` history a bit with all the merging, so a `rebase` might be needed.

Regarding the test, I am just creating a `DataFrame` with three columns equal to the integers `0, 1, ... , 9`, then creating another which adds an additional column of the same form with a duplicate column name.  The result of `corrwith` should be a `Series` of four ones (and should not generate an error).  Please let me know if there's a more straightforward way to simply check that the method doesn't err out.

```python
    def test_corrwith_dup_cols(self):
        # GH 21925
        df1 = pd.DataFrame(np.vstack([np.arange(10)] * 3).T)
        df2 = df1.copy()
        df2 = pd.concat((df2, df2[0]), axis=1)

        result = df1.corrwith(df2).values
        expected = np.ones(4)
        tm.assert_almost_equal(result, expected)
```

jreback commented:
can you rebase

pep8speaks commented:
Hello @dsaxton! Thanks for updating the PR.

 - There are no PEP8 issues in the file [`asv_bench/benchmarks/stat_ops.py`](https://github.com/dsaxton/pandas/blob/8b7436529547cbd9a0ffe93062aa7dfdb9a7401c/asv_bench/benchmarks/stat_ops.py) !

 - There are no PEP8 issues in the file [`pandas/core/frame.py`](https://github.com/dsaxton/pandas/blob/8b7436529547cbd9a0ffe93062aa7dfdb9a7401c/pandas/core/frame.py) !

 - There are no PEP8 issues in the file [`pandas/tests/frame/test_analytics.py`](https://github.com/dsaxton/pandas/blob/8b7436529547cbd9a0ffe93062aa7dfdb9a7401c/pandas/tests/frame/test_analytics.py) !

jreback reviewed (changes requested):
what is the perf diff with this (also add something with a lot of columns, like 100) to see what effect that has

jreback commented:
see #22684, a nice alternative

dsaxton commented:
@jreback  Should this PR still be implemented, or does #22684 make it obsolete?  (Since #22684 seems to be changing the implementation of `corr`, does it make sense to extend it to `corrwith`?)

dsaxton commented:
I'm not entirely sure how to show a perf diff, but here is something I put together that seems to show that the new implementation is faster than `corrwith` on `master` for Pearson's correlation, at least on my local machine. (Here we just have two scripts that only perform the part of the calculation that is different between `master` and this branch.  Each constructs two 100,000 x 100 `DataFrames` and then finds the correlation `Series` returned by `corrwith`.)

```
Daniels-MacBook-Pro:~ dsaxton$ cat perf1.py 
#!/usr/bin/env python3

import numpy as np
from pandas import DataFrame, Series

left = DataFrame(np.random.normal(size=(100000, 100)))
right = DataFrame(np.random.normal(size=(100000, 100)))

def c(x):
    return Series(x[0]).corr(Series(x[1]),
                  method="pearson")

correl = Series(map(c,
                    zip(left.values.T, right.values.T)),
                index=left.columns)
Daniels-MacBook-Pro:~ dsaxton$ cat perf2.py 
#!/usr/bin/env python3

import numpy as np
from pandas import DataFrame, Series

left = DataFrame(np.random.normal(size=(100000, 100)))
right = DataFrame(np.random.normal(size=(100000, 100)))

ldem = left - left.mean()
rdem = right - right.mean()

num = (ldem * rdem).sum()
dom = (left.count() - 1) * left.std() * right.std()

correl = num / dom
Daniels-MacBook-Pro:~ dsaxton$ time ./perf1.py 

real	0m2.025s
user	0m1.877s
sys	0m0.139s
Daniels-MacBook-Pro:~ dsaxton$ time ./perf2.py 

real	0m2.291s
user	0m2.128s
sys	0m0.740s
```

jreback commented:
pls look at the contributing docs on running the performance suite

the new impl does not scale with the number of columns - pls show this

dsaxton commented:
I think you're correct that the proposed implementation is not scaling.  Could it be due to `zip` and `map` not being as efficient as pure `numpy`?

```
Daniels-MacBook-Pro:asv_bench dsaxton$ asv continuous -f 1.1 upstream/master corrwith-dev -b stat_ops.Correlation
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Installing into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 8 total benchmarks (2 commits * 1 environments * 4 benchmarks)
[  0.00%] · For pandas commit hash e19b3f62:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  0.00%] ··· Running (stat_ops.Correlation.time_corr--)....
[ 12.50%] ··· stat_ops.Correlation.time_corr                                                                                                                         ok
[ 12.50%] ··· ========== ============
                method               
              ---------- ------------
               spearman    93.5±4ms  
               kendall     206±20ms  
               pearson    1.95±0.2ms 
              ========== ============

[ 25.00%] ··· stat_ops.Correlation.time_corrwith                                                                                                                     ok
[ 25.00%] ··· ========== ==========
                method             
              ---------- ----------
               spearman   39.6±5ms 
               kendall    35.0±2ms 
               pearson    35.9±3ms 
              ========== ==========

[ 37.50%] ··· stat_ops.Correlation.time_corrwith_cols                                                                                                                ok
[ 37.50%] ··· ========== ==========
                method             
              ---------- ----------
               spearman   115±6ms  
               kendall    19.5±1ms 
               pearson    36.4±2ms 
              ========== ==========

[ 50.00%] ··· stat_ops.Correlation.time_corrwith_rows                                                                                                                ok
[ 50.00%] ··· ========== ==========
                method             
              ---------- ----------
               spearman   731±20ms 
               kendall    458±30ms 
               pearson    237±7ms  
              ========== ==========

[ 50.00%] · For pandas commit hash 1a12c41d:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 50.00%] ··· Running (stat_ops.Correlation.time_corr--)....
[ 62.50%] ··· stat_ops.Correlation.time_corr                                                                                                                         ok
[ 62.50%] ··· ========== =============
                method                
              ---------- -------------
               spearman     85.9±5ms  
               kendall      200±4ms   
               pearson    2.03±0.07ms 
              ========== =============

[ 75.00%] ··· stat_ops.Correlation.time_corrwith                                                                                                                     ok
[ 75.00%] ··· ========== ============
                method               
              ---------- ------------
               spearman   19.8±0.5ms 
               kendall    19.4±0.3ms 
               pearson    19.6±0.5ms 
              ========== ============

[ 87.50%] ··· stat_ops.Correlation.time_corrwith_cols                                                                                                            failed
[ 87.50%] ··· ========== ========
                method           
              ---------- --------
               spearman   failed 
               kendall    failed 
               pearson    failed 
              ========== ========

[100.00%] ··· stat_ops.Correlation.time_corrwith_rows                                                                                                            failed
[100.00%] ··· ========== ========
                method           
              ---------- --------
               spearman   failed 
               kendall    failed 
               pearson    failed 
              ========== ========

       before           after         ratio
     [1a12c41d]       [e19b3f62]
+      19.8±0.5ms         39.6±5ms     2.00  stat_ops.Correlation.time_corrwith('spearman')
+      19.6±0.5ms         35.9±3ms     1.83  stat_ops.Correlation.time_corrwith('pearson')
+      19.4±0.3ms         35.0±2ms     1.80  stat_ops.Correlation.time_corrwith('kendall')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

jreback commented:
numpy is using a fully vectorizer soln

your impl is basically linear in the number of columns and thus doesn’t scale

dsaxton commented:
Got it, I will work on a numpy replacement for map

jreback commented:
nothing to do with map

the issue is a column by column impl is simply not scalable, that’s why it was written like this to begin with

dsaxton commented:
@jreback In the latest commit I'm retaining the current implementation of the Pearson calculation so there should be no difference in performance for this case, and only using the new approach for other correlation types. (While perhaps not perfectly scalable yet, I figure it will be a reasonable first effort.)  Also now that #22684 has been merged a nice side effect of this PR would be that `corrwith` can accept callable methods as well.

jreback commented:
can you merge master and see if you can get this to pass

dsaxton commented:
@TomAugspurger I added a commit with a merge (should I have rebased?  It seems many unrelated changes were added to the commit), a few fixes (the test failure was apparently caused by accidentally taking a transpose twice, and there was some error-catching code that wasn't accounting for callable methods) and an xfail added to the duplicate columns test

jreback commented:
you need to merge master. changes should only be your changes, not including other ones.

Commit:
Fix corrwith

* Remove incorrect error (didn't account for callables)
* Add xfail to duplicate columns test
* Fix transpose (was taken twice for Pearson)
* Remove inplace usage for dropna

dsaxton commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -466,6 +466,42 @@ def test_corrwith_mixed_dtypes(self):
+        expected = Series(np.ones(len(result)))
+        tm.assert_series_equal(result, expected)
+
+    def test_corrwith_invalid_method(self):
```
The relevant code here was actually removed because it didn't consider callable methods.  Should this test then be taken out as well?

Commit:
Remove test that's no longer correct

jreback commented:
can you merge master

jreback commented:
can you merge master and update to comments

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -370,8 +370,10 @@ Other Enhancements
-- :meth:`DataFrame.to_stata` and :class:`pandas.io.stata.StataWriter117` can write mixed sting columns to Stata strl format (:issue:`23633`)
-- :meth:`DataFrame.between_time` and :meth:`DataFrame.at_time` have gained the an ``axis`` parameter (:issue:`8839`)
+- :meth:`DataFrame.to_stata` and :class:` pandas.io.stata.StataWriter117` can write mixed sting columns to Stata strl format (:issue:`23633`)
+- :ref:`Series.resample` and :ref:`DataFrame.resample` have gained the :meth:`Resampler.quantile` (:issue:`15023`).
```
are these existing lines changed for some reason?

TomAugspurger replied:
We definitely don't want :ref: here. Maybe just a bad merge conflict that needs to be reverted.

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1260,7 +1262,8 @@ Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~
 
--
+- Added sphinx spelling extension, updated documentation on how to use the spell check (:issue:`21079`)
```
why was this changed? I think these are just old / rebase issues.

dsaxton replied:
This was a bad merge that I'm going to revert

dsaxton replied:
Looks like the revert didn't work as intended so I'm going to need to fix that as well

jreback commented on pandas/core/frame.py:
```diff
@@ -7021,53 +7026,77 @@ def cov(self, min_periods=None):
+            correl = num / dom
+
+        else:
+            def c(x):
```
I think its reasonable to simply do
```
def c(x):
    a, b = x
    return nanops.nancorr(x, y, method=method)
```
here as these are already aligned, and already arrays and don't need the extra checking

jreback commented on pandas/core/frame.py:
```diff
@@ -7021,53 +7026,77 @@ def cov(self, min_periods=None):
-        dom = (left.count() - 1) * left.std() * right.std()
+            correl = num / dom
+
+        else:
```
do elif method in ['.....], and add an else clause that raises (wrong method is passed)

jreback commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -466,6 +466,33 @@ def test_corrwith_mixed_dtypes(self):
+    @pytest.mark.xfail
+    def test_corrwith_dup_cols(self):
+        # GH 21925
+        df1 = pd.DataFrame(np.vstack([np.arange(10)] * 3).T)
```
can you add an example with an empty frame.

dsaxton replied:
A test to ensure that the output is an empty `Series` (with the proper index)?

TomAugspurger commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -466,6 +466,33 @@ def test_corrwith_mixed_dtypes(self):
         expected = pd.Series(data=corrs, index=['a', 'b'])
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.xfail
```
Perhaps I'm missing something, but why is this xfailed?

dsaxton replied:
I had xfailed this because master was giving an error when columns were duplicated (`ValueError: cannot reindex from a duplicate axis`)

TomAugspurger replied:
Could you rewrite the test to not have duplicate labels?

But I notice now that on master we do allow duplicates in `DataFrame.corrwith`

```python
In [6]: df = pd.DataFrame([[1, 2], [3, 4]], columns=[0, 0], index=[1, 1])

In [7]: df.corrwith(df)
Out[7]:
0    1.0
0    1.0
dtype: float64
```

so we may need to adjust the implementation to not regress on that. It'd be good to add some tests for that if they aren't already present.

dsaxton replied:
Hmm, I wonder why that code sample works but this one gives an error:

```
import numpy as np
import pandas as pd

df1 = pd.DataFrame(np.random.random(size=(10, 2)), columns=["a", "b"])
df2 = pd.DataFrame(np.random.random(size=(10, 2)), columns=["a", "a"])

df1.corrwith(df2)
```

dsaxton replied:
@TomAugspurger By duplicate labels do you mean the values within the `DataFrame` themselves (i.e., not the indices or column names)?

Commit:
Update corrwith

* Check for invalid method
* Do not cast arrays to Series in function c

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1260,7 +1261,7 @@ Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~
 
--
+- Improved clarity of `corrwith` docstring, added See Also entries for `corrwith` and `corr` (:issue:`22328`)
```
this is not needed

jreback commented on pandas/core/frame.py:
```diff
@@ -7021,53 +7026,79 @@ def cov(self, min_periods=None):
+                            index=left.columns)
+
+        else:
+            raise ValueError('Invalid method')
```
make this ``'Invalid method {} was passed, valid methods are: ......'``

jreback replied:
I *think* that is already done in the Series method, but certainly check

dsaxton replied:
Yes, seems you're correct (although the error message doesn't mention that callable methods are allowed; is this worth updating?)

Commit:
Fix error message

jreback commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -466,6 +466,33 @@ def test_corrwith_mixed_dtypes(self):
+
+        result = df1.corrwith(df2).values
+        expected = np.ones(4)
+        tm.assert_almost_equal(result, expected)
```
can you use ``assert_series_equal`` to make sure the indexes are correct

Commit:
Compare Series not arrays in test

jreback commented:
can you merge master, also update to handle duplicates

Commit:
Allow duplicate index labels

dsaxton commented on pandas/core/frame.py:
```diff
@@ -7102,7 +7102,9 @@ def c(x):
             raxis = 1 if axis == 0 else 0
             result_index = (this._get_axis(raxis).
                             union(other._get_axis(raxis)))
-            correl = correl.reindex(result_index)
```
Since the issue with duplicates occurs when we try to use the `reindex` method, my thought was to simply add in the missing values directly by appending to the existing `Series`

Commit:
Remove xfail from duplicate column test

Commit:
Use Series not pd.Series

jreback reviewed (changes requested):
minor comment. ping on green.

Commit:
Update corrwith

* Add comment for when drop is False
* Check if len(idx_diff) > 0
* Remove unnecessary string casting in error message

jreback commented:
lgtm. ping on green.

dsaxton commented:
@jreback Looks like it's passing.  Thanks for your help and patience!

jreback commented:
thanks @dsaxton
